cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -427,7 +437,44 @@ template <>
 hash_value_type __device__ inline SparkMurmurHash3_32<numeric::decimal128>::operator()(
   numeric::decimal128 const& key) const
 {
-  return this->compute<__int128_t>(key.value());
+  // Generates the Spark MurmurHash3 hash value, mimicking the conversion:
+  // java.math.BigDecimal.valueOf(unscaled_value, _scale).unscaledValue().toByteArray()
+  // https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L381
+  __int128_t const val               = key.value();
+  constexpr cudf::size_type key_size = sizeof(__int128_t);
+  std::byte const* data              = reinterpret_cast<std::byte const*>(&val);
+
+  // Small negative values start with 0xff..., small positive values start with 0x00...
+  bool const is_negative     = val < 0;
+  std::byte const zero_value = is_negative ? std::byte{0xff} : std::byte{0x00};
+
+  // Special cases for 0 and -1 which do not shorten correctly.
+  if (val == 0) { return this->compute(static_cast<uint8_t>(0)); }
+  if (val == static_cast<__int128>(-1)) { return this->compute(static_cast<uint8_t>(0xff)); }
+
+  // If the value can be represented with a shorter than 16-byte integer, the
+  // leading bytes of the little-endian value are truncated and are not hashed.
+  auto const reverse_begin = thrust::reverse_iterator(data + key_size);
+  auto const reverse_end   = thrust::reverse_iterator(data);
+  auto const first_nonzero_byte =
+    thrust::find_if_not(thrust::seq, reverse_begin, reverse_end, [zero_value](std::byte const& v) {
+      return v == zero_value;
+    }).base();
+  cudf::size_type length = thrust::distance(data, first_nonzero_byte);
```

Comment:
If we made `length` have a value of at least 1, would that eliminate the special cases that fail when shortening the values 0 and -1? It seems like it might work.
```suggestion
  cudf::size_type length = std::max(1, thrust::distance(data, first_nonzero_byte));
```

Reply:
This does solve the edge cases being explicitly handled, they have been removed